scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5082 @@
+        represented in floating point arithmetic, in which case this function
+        may be used to find the argument of the CDF for which the *logarithm*
+        of the resulting probability is `x`.
+
+        Due to the inherent spacing between floating point numbers, it is common
+        for there to be no value of the argument for which the log-CDF is close
+        to the provided `x`. In such cases, this function returns NaN.
+
+        Examples
+        --------
+        Instantiate a distribution with the desired parameters:
+
+        >>> import numpy as np
+        >>> from scipy import stats
+        >>> X = stats.Uniform(a=-0.5, b=0.5)
+
+        Evaluate the inverse log-CDF at the desired argument:
+
+        >>> X.ilogcdf(-0.25)
+        0.2788007830714034
+        >>> np.allclose(X.ilogcdf(-0.25), X.icdf(np.exp(-0.25)))
+        True
+
+        This function may return NaN when the spacing between floating point
+        values is too coarse to represent an argument for which the log-CDF is
+        close to the provided `x`, even when the function is theoretically
+        continuous within the domain of the argument.
+
+        >>> X.ilogcdf(-np.inf), X.ilogcdf(-1e10), X.ilogcdf(-30.)
+        (-0.5, nan, -0.4999999999998863)
+
+        """
+        return self._ilogcdf_dispatch(x, method=method, **self._parameters)
+
+    @_dispatch
+    def _ilogcdf_dispatch(self, x, method=None, **kwargs):
+        if self._overrides('_ilogcdf_formula'):
+            method = self._ilogcdf_formula
+        elif self._overrides('_ilogccdf_formula'):
+            method = self._ilogcdf_complementarity
+        else:
+            method = self._ilogcdf_inversion
+        return method
+
+    def _ilogcdf_formula(self, x, **kwargs):
+        raise NotImplementedError(self._not_implemented)
+
+    def _ilogcdf_complementarity(self, x, **kwargs):
+        return self._ilogccdf_dispatch(_log1mexp(x), **kwargs)
+
+    def _ilogcdf_inversion(self, x, **kwargs):
+        return self._solve_bounded(self._logcdf_dispatch, x, kwargs=kwargs)
+
+    @_set_invalid_nan
+    def icdf(self, x, *, method=None):
+        r"""Inverse cumulative distribution function
+
+        The inverse cumulative distribution function :math:`F^{-1}(x)` is the
+        argument :math:`y` for which the cumulative distribution function
+        :math:`F(y)` evaluates to :math:`x`.
+
+        .. math::
+
+            F^{-1}(x) = y \text{ such that } F(y) = x`
+
+        ``icdf`` accepts `x` for :math:`x \in [0, 1]`.
+
+        Parameters
+        ----------
+        x : Array
+            The argument of the inverse cumulative distribution function
+            (inverse CDF).
+        method : {None, 'formula', 'complementarity', 'inversion'}
+            The strategy used to evaluate the inverse CDF.
+            By default (``None``), the infrastructure chooses between the
+            following options.
+
+            - ``'formula'``: use a formula for the inverse CDF itself
```

Comment:
I don't understand the comment. If the user specifies `formula`, they are specifying that the `_icdf_formula` should be used. If it is not overridden by the author of the distribution, a `NotImplementedError` is raised. After each block of these options, there is a standard comment that not all method options will be available for all distributions.

Reply:
My comment is about `method=None`. The resolution order would be good to document, either here or somewhere central. If needs be, the documenting may qualify that it's the current resolution order and it may change in the future.

Reply:
For now, I';ve changed this to "between the following options, listed in order of precedence."

Describing precisely what happens is going to be a little nuanced. For example, the logic of `ccdf_dispatch` is:

```python3
    def _ccdf_dispatch(self, x, method=None, **kwargs):
        if self._overrides('_ccdf_formula'):
            method = self._ccdf_formula
        elif self._overrides('_logccdf_formula'):
            method = self._ccdf_logexp
        elif self._tol is _null and self._overrides('_cdf_formula'):
            method = self._ccdf_complementarity
        else:
            method = self._ccdf_quadrature
        return method
```

Compare against the language:

> By default (``None``), the infrastructure chooses between the following options, listed in order of precedence.
> - ``'formula'``: use a formula for the CCDF itself
> - ``'logexp'``: evaluate the logarithm of the CCDF directly and exponentiate
> - ``'complementarity'``: evaluate the CDF and take the complement
> - ``'quadrature'``: numerically integrate the PDF

So this really means:
- use a formula for the CCDF itself if implemented
- if a formula for the logarithm of the CCDF is implement, evaluate the logarithm of the CCDF directly and exponentiate
- if a formula for the CDF is implemented and no tolerance is specified, evaluate the CDF and take the complement
- otherwise, fall back to quadrature.

There are other layers that would be easy to add and probably should be added, since they take essentially no time to resolve, add little complexity to the code, and can avoid bug reports about inaccurate this and that. For instance, if a formula for the `logcdf` (not `logccdf`) is available, it would make sense to take the log-complement and exponentiate the result before directly computing the complement of the CDF.

Also note that the tolerance thing is only notional right now. What that really is supposed to mean is that if taking 1 - CDF would make the resulting precision less than the tolerance, we'd skip over that and integrate directly.

So as simple as the code is, the precise logic is a bit involved, and I didn't want to get bogged down at this level in a first pass of the documentation, and I'd rather omit information than purport to include it and be inaccurate. Since the logic is probably going to change a bit, I was hoping to leave it out initially and add it somewhere when it settles.

I did prepare these diagrams and hoped to include revisions at some point:
![image](https://github.com/mdhaber/scipy/assets/6570539/ecd307a2-ad8d-47b0-ba29-94afab99480a)

Reply:
We discussed leaving this alone.